File: api/app/models/parcelle.py
"""Types métier de la parcelle cacaoyère et de ses captures terrain.

Deux familles cohabitent, comme dans ``app/models/session.py`` :

* les **types de domaine** (``dataclass(frozen=True)``) — immuables, sans dépendance
  framework, manipulés par les services et le dépôt ;
* les **schémas d'API** (Pydantic v2) — validation des entrées et des sorties HTTP.

L'immuabilité n'est pas un ornement : une capture qui traverse plusieurs couches
(recevabilité, écriture disque, persistance) ne doit pas pouvoir être modifiée en
route par surprise.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field, model_validator

from app.services.geometrie import dans_cote_ivoire, superficie_ha

# Bornes de plausibilité d'une parcelle cacaoyère ivoirienne. En dessous de 0,1 ha le
# tracé est une erreur de manipulation ; au-delà de 50 ha ce n'est plus une parcelle
# de producteur mais un domaine, qui relève d'un autre traitement.
SUPERFICIE_MIN_HA = 0.1
SUPERFICIE_MAX_HA = 50.0

# Un anneau exploitable demande au moins quatre sommets distincts (un triangle fermé
# par le parcours GPS est presque toujours un tracé abandonné).
POINTS_MIN_POLYGONE = 4

# Plafond d'images par capture, aligné sur l'échantillonnage vidéo du navigateur
# (1 image / 2 s). Douze vues suffisent à un constat et bornent le téléversement.
IMAGES_MAX_PAR_CAPTURE = 12

# Bornes du score de netteté (variance du laplacien) calculé par le navigateur.
SCORE_NETTETE_MIN = 0.0
SCORE_NETTETE_MAX = 100_000.0

NOM_MAX = 120
LOCALITE_MAX = 120


class TypeGeometrie(str, Enum):
    """Nature géométrique d'une parcelle."""

    POINT = "point"
    POLYGONE = "polygone"


class SourceGeometrie(str, Enum):
    """Provenance de la géométrie."""

    PARCOURS_GPS = "parcours_gps"
    SAISIE_MANUELLE = "saisie_manuelle"


class Modalite(str, Enum):
    """Modalité de capture terrain."""

    PHOTOS = "photos"
    VIDEO = "video"
    PARCOURS = "parcours"
    PARCOURS_VIDEO = "parcours_video"


class MotifRecevabilite(str, Enum):
    """Motif du verdict de recevabilité d'une image."""

    OK = "ok"
    FLOU = "flou"
    SOUS_EXPOSE = "sous_expose"
    SUR_EXPOSE = "sur_expose"
    TROP_PETITE = "trop_petite"
    FORMAT_REFUSE = "format_refuse"


@dataclass(frozen=True)
class Coordonnee:
    """Point géographique horodaté, tel que rendu par le navigateur."""

    latitude: float
    longitude: float
    precision_m: float | None = None
    horodatage: datetime | None = None


@dataclass(frozen=True)
class Geometrie:
    """Géométrie d'une parcelle : un point, ou un anneau et sa superficie."""

    type: TypeGeometrie
    points: tuple[Coordonnee, ...]
    source: SourceGeometrie
    superficie_ha: float | None = None

    @classmethod
    def depuis_points(cls, points: tuple[Coordonnee, ...], source: SourceGeometrie) -> Geometrie:
        """Construit une géométrie et calcule sa superficie si c'est un anneau.

        Args:
            points: Sommets relevés, dans l'ordre du parcours.
            source: Provenance des points.

        Returns:
            Une géométrie de type ``POINT`` si un seul point, ``POLYGONE`` sinon.
        """
        if len(points) < POINTS_MIN_POLYGONE:
            return cls(type=TypeGeometrie.POINT, points=points, source=source)
        surface = superficie_ha([(p.latitude, p.longitude) for p in points])
        return cls(
            type=TypeGeometrie.POLYGONE,
            points=points,
            source=source,
            superficie_ha=surface,
        )


@dataclass(frozen=True)
class Recevabilite:
    """Verdict de recevabilité d'une image, avec conseil de reprise."""

    recevable: bool
    motif: MotifRecevabilite
    conseil: str
    score_nettete: float


@dataclass(frozen=True)
class Image:
    """Image persistée : identifiée par l'empreinte de son contenu."""

    empreinte_sha256: str
    largeur: int
    hauteur: int
    recevabilite: Recevabilite
    coordonnee: Coordonnee | None = None


@dataclass(frozen=True)
class Capture:
    """Une session de capture terrain rattachée à une parcelle."""

    identifiant: str
    parcelle: str
    proprietaire: str
    modalite: Modalite
    cree_le: datetime
    images: tuple[Image, ...] = ()
    trace: tuple[Coordonnee, ...] = ()


@dataclass(frozen=True)
class Parcelle:
    """Une parcelle cacaoyère déclarée par un producteur."""

    identifiant: str
    proprietaire: str
    nom: str
    localite: str
    direction_regionale: str
    cree_le: datetime
    maj_le: datetime
    geometrie: Geometrie | None = None


# --------------------------------------------------------------- schémas d'API


class CoordonneeRequest(BaseModel):
    """Point géographique reçu du navigateur."""

    latitude: float = Field(ge=-90.0, le=90.0)
    longitude: float = Field(ge=-180.0, le=180.0)
    precision_m: float | None = Field(default=None, ge=0.0, le=10_000.0)
    horodatage: datetime | None = None


class ImageRequest(BaseModel):
    """Image téléversée, encodée en base64.

    L'encodage base64 évite la dépendance ``python-multipart``, conformément au choix
    déjà retenu pour la console de curation (``app/curation/models.py``).

    Les métriques ``score_nettete`` et ``luminance_moyenne`` sont calculées par le
    navigateur, qui possède les pixels décodés. Le serveur les borne sans leur faire
    confiance : ce sont des indications de confort, pas un contrôle de sécurité.
    """

    contenu_base64: str = Field(min_length=1)
    largeur: int = Field(ge=1, le=20_000)
    hauteur: int = Field(ge=1, le=20_000)
    score_nettete: float = Field(ge=SCORE_NETTETE_MIN, le=SCORE_NETTETE_MAX)
    luminance_moyenne: float = Field(default=128.0, ge=0.0, le=255.0)
    coordonnee: CoordonneeRequest | None = None


class CaptureRequest(BaseModel):
    """Dépôt d'une capture : des images, une trace, ou les deux."""

    modalite: Modalite
    images: list[ImageRequest] = Field(default_factory=list, max_length=IMAGES_MAX_PAR_CAPTURE)
    trace: list[CoordonneeRequest] = Field(default_factory=list, max_length=2_000)

    @model_validator(mode="after")
    def _au_moins_une_donnee(self) -> CaptureRequest:
        """Refuse une capture qui n'apporte ni image ni trace."""
        if not self.images and not self.trace:
            raise ValueError("Une capture doit contenir au moins une image ou une trace.")
        return self


class GeometrieRequest(BaseModel):
    """Enregistrement d'une géométrie de parcelle (parcours ou saisie)."""

    points: list[CoordonneeRequest] = Field(min_length=1, max_length=2_000)
    source: SourceGeometrie = SourceGeometrie.PARCOURS_GPS


class CreerParcelleRequest(BaseModel):
    """Création d'une parcelle."""

    nom: str = Field(min_length=1, max_length=NOM_MAX)
    localite: str = Field(min_length=1, max_length=LOCALITE_MAX)


class RecevabiliteReponse(BaseModel):
    """Verdict de recevabilité exposé au client."""

    recevable: bool
    motif: MotifRecevabilite
    conseil: str
    score_nettete: float


class ImageReponse(BaseModel):
    """Image telle que persistée, exposée au client."""

    empreinte_sha256: str
    largeur: int
    hauteur: int
    recevabilite: RecevabiliteReponse
    coordonnee: CoordonneeRequest | None = None


class GeometrieReponse(BaseModel):
    """Géométrie exposée au client, superficie comprise."""

    type: TypeGeometrie
    points: list[CoordonneeRequest]
    source: SourceGeometrie
    superficie_ha: float | None = None


class ParcelleReponse(BaseModel):
    """Parcelle exposée au client."""

    identifiant: str
    nom: str
    localite: str
    direction_regionale: str
    cree_le: datetime
    maj_le: datetime
    geometrie: GeometrieReponse | None = None


class CaptureReponse(BaseModel):
    """Capture exposée au client."""

    identifiant: str
    parcelle: str
    modalite: Modalite
    cree_le: datetime
    images: list[ImageReponse] = Field(default_factory=list)
    trace: list[CoordonneeRequest] = Field(default_factory=list)


def point_ivoirien(point: CoordonneeRequest) -> bool:
    """Indique si un point reçu tombe dans l'enveloppe de la Côte d'Ivoire."""
    return dans_cote_ivoire(point.latitude, point.longitude)

File: web/src/domain/parcelle.js
// Couche DOMAINE — parcelle cacaoyère et captures terrain (V3, chantier C1).
// Vocabulaire, plafonds et règles pures : ni DOM, ni fetch, ni canvas, ni capteur.
// Les bornes sont alignées sur celles de l'API (api/app/models/parcelle.py) : le
// navigateur refuse AVANT de consommer la bande passante d'un réseau mobile, mais
// c'est bien le serveur qui fait autorité — ces constantes ne sont qu'un miroir.

/** Les quatre modalités de capture terrain (miroir de `Modalite` côté API). */
export const Modalite = Object.freeze({
  PHOTOS: "photos",
  VIDEO: "video",
  PARCOURS: "parcours",
  PARCOURS_VIDEO: "parcours_video",
});

/** Provenance d'une géométrie (miroir de `SourceGeometrie` côté API). */
export const SourceGeometrie = Object.freeze({
  PARCOURS_GPS: "parcours_gps",
  SAISIE_MANUELLE: "saisie_manuelle",
});

/** Libellés des modalités, pour l'affichage. */
export const LIBELLE_MODALITE = Object.freeze({
  [Modalite.PHOTOS]: "Photos",
  [Modalite.VIDEO]: "Vidéo",
  [Modalite.PARCOURS]: "Parcours",
  [Modalite.PARCOURS_VIDEO]: "Parcours + vidéo",
});

// --- Plafonds de capture -----------------------------------------------------

// Douze vues suffisent à un constat, et bornent le téléversement (miroir de
// IMAGES_MAX_PAR_CAPTURE côté API).
export const IMAGES_MAX = 12;

// Une image toutes les deux secondes de vidéo : au-delà, deux images voisines
// montrent la même chose.
export const INTERVALLE_ECHANTILLON_S = 2;

// Redimensionnement OBLIGATOIRE avant envoi. Une photo de téléphone moderne fait
// 4000 px de large ; encodée en base64 elle pulvériserait à elle seule le plafond
// de 8 Mio du corps de requête. À 1024 px et qualité 0,85, une vue pèse ~200 Ko,
// et douze vues tiennent largement dans le plafond.
export const LARGEUR_MAX_PX = 1024;
export const QUALITE_JPEG = 0.85;

// Un point qui n'a pas bougé de 5 m n'apporte aucune information sur le contour.
export const DISTANCE_MIN_M = 5;

// Au-delà de 50 m d'incertitude, le point dégrade le tracé au lieu de l'enrichir.
export const PRECISION_MAX_M = 50;

// Plafond de précision déclarable à l'API (Field(le=10_000) côté Pydantic) : une
// valeur plus grande ferait échouer toute la capture en 422, on l'omet plutôt.
export const PRECISION_DECLARABLE_MAX_M = 10_000;

// Miroirs des bornes Pydantic de l'API.
export const TRACE_MAX_POINTS = 2_000;
export const POINTS_MIN_POLYGONE = 4;
export const SCORE_NETTETE_MAX = 100_000;
export const NOM_MAX = 120;
export const LOCALITE_MAX = 120;

// --- Recevabilité ------------------------------------------------------------

/** Libellé court du motif de recevabilité rendu par le serveur. */
export const LIBELLE_MOTIF = Object.freeze({
  ok: "Photo exploitable",
  flou: "Photo floue",
  sous_expose: "Photo trop sombre",
  sur_expose: "Photo éblouie",
  trop_petite: "Photo trop petite",
  format_refuse: "Fichier non reconnu",
});

// Conseils de repli : le serveur en renvoie toujours un, ceux-ci ne servent qu'au
// cas — anormal — d'une réponse sans conseil. Français simple, pas de jargon.
const CONSEIL_DEFAUT = Object.freeze({
  ok: "Photo exploitable.",
  flou: "Approchez-vous, tenez le téléphone immobile, et refaites la photo.",
  sous_expose: "Sortez de l'ombre, puis refaites la photo.",
  sur_expose: "Tournez-vous dos au soleil, puis refaites la photo.",
  trop_petite: "Utilisez l'appareil photo du téléphone, pas une capture d'écran.",
  format_refuse: "Envoyez une photo prise avec le téléphone (JPEG ou PNG).",
});

/**
 * Libellé d'un motif de recevabilité, sûr même si le serveur en ajoute un nouveau.
 * @param {string} motif
 */
export function libelleMotif(motif) {
  return LIBELLE_MOTIF[motif] || "Photo non retenue";
}

/**
 * Conseil de reprise à afficher : celui du serveur d'abord, un repli sinon.
 * @param {{motif: string, conseil: string}} recevabilite
 */
export function conseilRecevabilite(recevabilite) {
  const conseil = recevabilite && typeof recevabilite.conseil === "string" ? recevabilite.conseil : "";
  return conseil.trim() || CONSEIL_DEFAUT[recevabilite?.motif] || "";
}

// --- Géométrie ---------------------------------------------------------------

// Rayon moyen de la Terre (IUGG), en mètres — même valeur que services/geometrie.py.
const RAYON_TERRE_M = 6_371_008.8;

/**
 * Distance entre deux points, en mètres, par projection équirectangulaire locale.
 * À l'échelle d'une parcelle l'erreur est négligeable, et l'on évite l'absurdité
 * d'une comparaison en degrés bruts (un degré de longitude ne vaut pas un degré
 * de latitude à 6° de latitude nord).
 * @param {{latitude: number, longitude: number}} a
 * @param {{latitude: number, longitude: number}} b
 * @returns {number} distance en mètres
 */
export function distanceM(a, b) {
  const facteur = (Math.PI / 180) * RAYON_TERRE_M;
  const latMoyenne = ((a.latitude + b.latitude) / 2) * (Math.PI / 180);
  const dx = (b.longitude - a.longitude) * facteur * Math.cos(latMoyenne);
  const dy = (b.latitude - a.latitude) * facteur;
  return Math.sqrt(dx * dx + dy * dy);
}

/**
 * Un point apporte-t-il de l'information par rapport au précédent retenu ?
 * @param {{latitude: number, longitude: number}|null|undefined} dernier
 * @param {{latitude: number, longitude: number}} candidat
 */
export function pointApporteInfo(dernier, candidat) {
  if (!dernier) return true;
  return distanceM(dernier, candidat) >= DISTANCE_MIN_M;
}

// --- Normalisation des réponses ----------------------------------------------

const nombre = (valeur, defaut = 0) => (typeof valeur === "number" && isFinite(valeur) ? valeur : defaut);
const texte = (valeur) => (typeof valeur === "string" ? valeur : "");

/** Normalise un point géographique rendu par l'API. */
export function versCoordonnee(brut) {
  if (!brut || typeof brut !== "object") return null;
  return Object.freeze({
    latitude: nombre(brut.latitude),
    longitude: nombre(brut.longitude),
    precisionM: typeof brut.precision_m === "number" ? brut.precision_m : null,
    horodatage: texte(brut.horodatage),
  });
}

/** Normalise la géométrie d'une parcelle (contour et superficie calculée). */
export function versGeometrie(brut) {
  if (!brut || typeof brut !== "object") return null;
  const points = Array.isArray(brut.points) ? brut.points.map(versCoordonnee).filter(Boolean) : [];
  return Object.freeze({
    type: texte(brut.type),
    source: texte(brut.source),
    points,
    superficieHa: typeof brut.superficie_ha === "number" ? brut.superficie_ha : null,
  });
}

/** Normalise une parcelle rendue par l'API. */
export function versParcelle(brut) {
  return Object.freeze({
    identifiant: texte(brut?.identifiant),
    nom: texte(brut?.nom),
    localite: texte(brut?.localite),
    directionRegionale: texte(brut?.direction_regionale),
    creeLe: texte(brut?.cree_le),
    majLe: texte(brut?.maj_le),
    geometrie: versGeometrie(brut?.geometrie),
  });
}

/** Normalise le verdict de recevabilité d'une image. */
export function versRecevabilite(brut) {
  return Object.freeze({
    recevable: Boolean(brut?.recevable),
    motif: texte(brut?.motif),
    conseil: texte(brut?.conseil),
    scoreNettete: nombre(brut?.score_nettete),
  });
}

/** Normalise une image persistée d'une capture. */
export function versImageCapture(brut) {
  return Object.freeze({
    empreinte: texte(brut?.empreinte_sha256),
    largeur: nombre(brut?.largeur),
    hauteur: nombre(brut?.hauteur),
    recevabilite: versRecevabilite(brut?.recevabilite),
    coordonnee: versCoordonnee(brut?.coordonnee),
  });
}

/** Normalise une capture rendue par l'API. */
export function versCapture(brut) {
  return Object.freeze({
    identifiant: texte(brut?.identifiant),
    parcelle: texte(brut?.parcelle),
    modalite: texte(brut?.modalite),
    creeLe: texte(brut?.cree_le),
    images: Array.isArray(brut?.images) ? brut.images.map(versImageCapture) : [],
    trace: Array.isArray(brut?.trace) ? brut.trace.map(versCoordonnee).filter(Boolean) : [],
  });
}
